### DISCUSSION TITLE
How to stop multiple re-rerenders and isolate onChange?

### DISCUSSION THREAD

**Discussion by drydenwilliams** [react-hook-form/react-hook-form#7611] [Q&A] [answered]
I would like to know how to stop every input/component from re-rendering when an input changes?

I would like to render a form with the corresponding data (from my API) like this: 

```
{
  a: 0,
  b: 0,
  c: true
}
```

and wondered how I could split out the form components but stop them from re-rendering if another component is updated?

[Codesandbox example](https://codesandbox.io/s/re-rendering-independent-input-forked-3zkvo?file=/src/Content.js)

Currently when you change one component, it re-renders all of them, which is ok for this example but the object in real life is quite big. 

Thanks for your help in advance 😄

**Comment by Moshyfawn** [comment]
- Consider using [`useWatch`](https://react-hook-form.com/api/usewatch) instead of [`watch`](https://react-hook-form.com/api/useform/watch) to localize rerenders at the component level where the value actually needs to watched.
- The use of [`useFormState`](https://react-hook-form.com/api/useformstate) instead of [`useFormContext`](https://react-hook-form.com/api/useformcontext) to get only the form's state instead of all it's methods.
- Use [`getValues`](https://react-hook-form.com/api/useform/getvalues) to retrieve some form value instead of [`watch`](https://react-hook-form.com/api/useform/watch) to avoid subscribing (aka. rerendering) to form values.
- Use [`reset`](https://react-hook-form.com/api/useform/reset) when you need to manually set most of the forms values instead of `setValue` which sets them one by one.
- Consider uncontrolled inputs via [`register`](https://react-hook-form.com/api/useform/register) instead of [`Controller`](https://react-hook-form.com/api/usecontroller/controller) / [`useController`](https://react-hook-form.com/api/usecontroller) to avoid updating the forms state

Also, see[ this FormPrivder perfomance section](https://react-hook-form.com/advanced-usage#FormProviderPerformance) of RHF docs

**Reply by drydenwilliams** [reply]
Thanks @Moshyfawn, I've added `useWatch` but it doesn't seem to stop the re-render? Any chance you could edit the codesandbox and show me?

**Reply by drydenwilliams** [reply]
I updated the example for a more "read world" example of what I'm trying to achieve, I hope that's helpful: 

https://codesandbox.io/s/re-rendering-independent-input-forked-3zkvo?file=/src/ThirdFormSection.js

It's basically a big form, but I would like to split it up into different sections so I can conditionally render parts based on other values in the form... without lots of re-rendering.

**Comment by Moshyfawn** [comment, marked as answer]
Alright, see [this CSB](https://codesandbox.io/s/dynamic-radio-example-forked-et0wi?file=/src/content/FirstFormSection.tsx) for some comments and rerenders data.
_Sry to use another CSB; caused an infinite rerender in a fork of your CSB by accident, so it didn't open for while_

**Reply by Moshyfawn** [reply]
It also emulated getting remote data, hence a rerender in a sec after you load CSB

**Reply by drydenwilliams** [reply]
Oh my! It's wonderful, thanks so much for this; really appreciate it and so many cool things I didn't know about like: `useController`, `useWatch` and `useFormContext` 👍 

I think/hope this will be a wonderful example for a lot of people making their forms even more performant!
